text/0000-coroutines.md

```diff
@@ -0,0 +1,485 @@
+
+```rust
+impl Drop for CoroClosure1234 {
+    fn drop(&mut self) {
+        match (self.state) {
+            1 => { /* clean-up locals alive at yield point 1 */ },
+            2 => { /* clean-up locals alive at yield point 2 */ }
+            ...
+        }
+    }
+}
+
+```
+
+### Fn* traits
+Coroutines shall implement the `FnMut` trait:
+- they cannot just implement `Fn`, because at the very least they need to modify the field which keeps
+  track of the current state,
+- they cannot implement `FnOnce`, because then the environment would get destroyed after the first yield.
+
+### Translation
+- Most rustc passes stay the same as for regular closures.
+- Type inference and checking passes are modified to take into account the new rules concerning types of 
+  arguments and return values.
+- Borrow checking must ensure that no borrows are live across yield points.
+- Code generation is modified as follows:
+    - A 'state' variable is added into the coroutine environment.
+    - Local variables whose lifetime straddles any yield point are hoisted into the coroutine environment.  
+    - A "master switch" is added at the top of the function to transfer control to the correct location 
+      according to the current state.
+    - **yield** expressions are translated as 
+        ```rust
+        self.state = <N>; 
+        return Yield(expr);
+        ```
+    - **return** statements and the tail expression are translated as 
+        ```rust
+        self.state = -1; 
+        return Return(<expr>);
+        ```
+
+Putting all this together, the `coro2` example above would be translated into something like this:
+```rust
+struct CoroClosure1234 {
+    state : int;
+    i : int;
+    // closed-over variables of the containing function (upvars) also go here
+}
+
+impl FnMut<(A1, A2, A3)> for CoroClosure1234 {
+    type Output = CoResult<(typeof y1, y2), (typeof result)>;
+
+    fn call_mut(&mut self, a1:A1, a2:A2, a3: A3) -> Self::Output {
+        // The body of a coroutine is not expressible in plain Rust,
+        // so I am using a MIR-like notation here.
+        entry: {
+            switchInt (self.state) -> [0:state_0, 1:state_1, 2:state_2, otherwise: invalid];
+        } 
+
+        state_0: {
+            ...
+            self.i = 0;
+        }
+
+        bb1: {
+            if (self.i < 10) -> [true: bb2:, false: end];
+        }
+
+        bb2: {
+            self.state = 1;
+            return Yield(y1);
+        }
+
+        state_1: {
+            let (b1,b2,b3) = (a1, a2, a3);
+            ...
+            self.state = 2;
+            return Yield(y2);
+        }
+
+        state_2: {
+            let (c1,c2,c3) = (a1, a2, a3);
+            ...
+            self.i += 1;
+            goto -> bb1;
+        }
+
+        ...
+        end: {
+            self.state = -1;
+            return Return(result);
+        }
+
+        invalid: {
+            ...
+            std::rt::begin_panic(const "invalid state!")
```

Comment:
It also seems to me that a somewhat builder-like pattern could be used to avoid this failure case. Namely, if coroutine instead was:

```
enum CoResult<Y,R,C> {
    Yield<Y, C>,
    Return<R>
}
```

and the coroutine closure was required to be passed around via value, then you’d evade the issue with cleanups on panic within a coroutine and also would never be able to “call” coroutine again after it has already returned.